zip-0315.rst

```diff
@@ -4,6 +4,451 @@
+how long until they expire, and their balances as described in `Reporting of balances`_.
+
+
+Transaction creation
+--------------------
+
+Anchor selection
+''''''''''''''''
+
+TODO: choose anchors at the trusted confirmation depth. This avoids leaking information
+about whether or not we spent trusted notes.
+
+
+Rationale for anchor selection
+''''''''''''''''''''''''''''''
+
+TODO
+
+
+Linkability of transactions or addresses
+----------------------------------------
+
+TODO: dusting attack mitigation
+
+
+Network-layer privacy
+---------------------
+
+
+Viewing keys
+------------
+
+What they are supposed to reveal; see ZIP 310 for Sapling (needs updating for
+Orchard). https://github.com/zcash/zips/issues/606
+
+
+
+Allowed transfers
+-----------------
+
+* Sprout -> transparent or Sapling
+* Sapling -> transparent or Sapling or Orchard
+* Orchard -> transparent or Sapling or Orchard
+* if auto-shielding is off:
+  *  transparent -> transparent or Sapling or Orchard
+* if auto-shielding is on:
+  *  transparent -> internal Orchard or Sapling
+
+Note: wallets MAY further restrict the set of transfers they perform.
+
+
+Auto-shielding
+--------------
+
+Wallets SHOULD NOT spend funds from a transparent address to an external address,
+unless the user gives explicit consent for this on a per-transaction basis.
+
+In order to support this policy, wallets SHOULD implement a system of auto-shielding
+with the following characteristics.
+
+
+If auto-shielding functionality is available in a wallet, then users MUST be able
+to explicitly consent to one of the following possibilities:
+
+* auto-shielding is always on;
+* auto-shielding is always off;
+* the user specifies a policy...
+
+Auto-shielding MUST be one of:
+
+* "must opt in or out" (zcashd will do this -- i.e. refuse to start unless the option
+  is configured), or
+* always on.
+
+
+Auto-migration
+--------------
+
+
+Information leakage for transfers between pools
+-----------------------------------------------
+
+
+If no auto-migration, if you can satisfy a transfer request to Sapling from your
+Sapling funds, do so.
+
+The user's consent is needed to reveal amounts. Therefore, there should be
+per-transaction opt-in for any amount-revealing transfer.
+
+* there may be a compatibility issue for amount-revealing cross-pool txns that were
+  previously allowed without opt-in
+
+Don't automatically combine funds across pools to satisfy a transfer (since that
+could reveal the total funds in some pool).
+
+In order to maintain the integrity of IVK guarantees, wallets should not generate
+unified addresses that contain internal receivers, nor expose internal receivers
+(such as those used for auto-shielding and change outputs) in any way.
+
+Open questions:
+
+* should there be an auto-migration option from Sapling to Orchard?
+
+# str4d notes
+
+If we want to have both automatic and opportunistic shielding, and keep the two
```

Comment:
Opportunistic Shielded understood as "tailgating a z2z transaction and shield funds at the same time" Should be discouraged. 

Suppose a user withdraws a significant amount of funds from a centralized Transparent-only exchange to its wallet. That same user forgets about that transaction and sends a payment to a third party on a different matter, for example a coffee. The wallet proceeds to opportunistically shield the transparent funds by hoping to that coffee purchase. The user inadvertently ends up tying a shielding operation to a mundane purchase. The merchant knows who paid in person and knows the transaction ID and can see the transparent part of the transaction. 

Shielding transactions should always be on their own and never be tied up to other transactions.

Reply:
> Opportunistic Shielded understood as "tailgating a z2z transaction and shield funds at the same time" Should be discouraged.

I agree, and we should clarify what we mean here.

I intend Opportunistic Shielding to be interpreted the same way we do in `zcashd`: if a wallet is willing to select from both transparent and shielded, and there is not sufficient shielded funds to fill the transaction (so transparent funds are _required_, and the transaction _must_ be t2z), and the user is willing to allow visible senders, then the wallet should take the opportunity to shield as much as it can.

This doesn't affect the current mobile SDK APIs, which explicitly _don't_ select transparent funds when creating transactions. They would not ever run Opportunistic Shielding, and instead only run Automatic Shielding via a separate process.

Reply:
Cool that's great to know!

Something that comes to mind is that even though the user might be trying to seize the opportunity, a transparent balance as the one I described may have the same side effect even if the user has agreed to opportunistically shield. 


I'm specifically thinking of a scenario like this one. Let's say:

precondition A: the user withdrew a non-trivial amount of ZEC to a T-address from a centralized exchange that only supports T.

Precondition B: has other trivial transparent funds, that arrived after Precondition A funds.

Precondition C: funds in A and B are Confirmed and spendable

Assumption: Users might more likely remember recent transactions before older ones.

Intent:  User is trying to fulfill a payment and the whole amount can't be met with the available shielded funds. And uses Opportunistic Shielding to cover for the reminding funds for the transaction.

User expectation: the smallest amount will be used to complete the transaction amount I choose convenience over higher privacy because the amount I need is trivial to me. 

result: the wallet will try to spend the oldest funds first, picking the non-trivial UTXO and possibly disclosing a balance to a third party 

Is this scenario feasible? Am I missing other aspects that make it impossible?

If it is possible, should wallets warn about the UTXOs involved so that users can be aware of what could possibly be disclosed by the opportunistic shielding transaction?